### PR TITLE
applet.interface.jtag_openocd: implement remote sleep

### PR DESCRIPTION
**Pull request by neuschaefer**
Note: This PR has prerequesites:

- [x] [A change to the OpenOCD remote_bitbang protocol](https://review.openocd.org/c/openocd/+/8191), which is ~~currently under review~~ merged

And a TODO list:
- [x] clean up according to the protocol change (once reviewed)
- [ ] improve documentation, as previously attempted in https://github.com/GlasgowEmbedded/glasgow/pull/522
- [ ] port changes to swd-openocd as well

### CODE REVIEW

**Comment by neuschaefer** [comment]
@jeremyherbert this is probably of interest to you

**Comment by jeremyherbert** [comment]
Great work! Fixes https://github.com/GlasgowEmbedded/glasgow/issues/303 from the looks of it

File: software/glasgow/applet/interface/jtag_openocd/__init__.py [hunk 1]
```diff
@@ -12,11 +12,13 @@
...
 
 class JTAGOpenOCDSubtarget(Elaboratable):
-    def __init__(self, pads, out_fifo, in_fifo, period_cyc):
+    def __init__(self, pads, out_fifo, in_fifo, period_cyc, us_cyc):
         self.pads       = pads
         self.out_fifo   = out_fifo
         self.in_fifo    = in_fifo
         self.period_cyc = period_cyc
+        self.us_cyc     = us_cyc
+        self.ms_cyc     = us_cyc * 1000
```
Comment (whitequark) [inline software/glasgow/applet/interface/jtag_openocd/__init__.py:21]: I'd rather not have this as a variable since then you can't have `ms_cyc` and `us_cyc` ever disagree.

Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py]: fixed

File: software/glasgow/applet/interface/jtag_openocd/__init__.py [hunk 2]
```diff
@@ -118,6 +129,7 @@ def build(self, target, args):
             out_fifo=iface.get_out_fifo(),
             in_fifo=iface.get_in_fifo(),
             period_cyc=int(target.sys_clk_freq // (args.frequency * 1000)),
+            us_cyc=int(target.sys_clk_freq // 1000_000),
```
Comment (whitequark) [inline software/glasgow/applet/interface/jtag_openocd/__init__.py:132]:
```suggestion
            us_cyc=int(target.sys_clk_freq // 1_000_000),
```

File: software/glasgow/applet/interface/jtag_openocd/__init__.py [hunk 3]
```diff
@@ -64,13 +69,19 @@ def elaborate(self, platform):
...
                     # remote_bitbang_quit(void)
                     with m.Case(*b"Q"):
                         pass
                     with m.Default():
+                        # For unknown commands, deliberately hang by requesting
+                        # no more FIFO elements, and skip the delay
                         m.d.comb += out_fifo.r_en.eq(0)
-                with m.If(out_fifo.r_en):
-                    m.d.sync += timer.eq(self.period_cyc - 1)
+                        m.d.sync += timer.eq(self.period_cyc - 1)
```
Comment (whitequark) [inline software/glasgow/applet/interface/jtag_openocd/__init__.py:84]: Why is this necessary? The FIFO will not progress anyway.

File: software/glasgow/applet/interface/jtag_openocd/__init__.py [hunk 4]
```diff
@@ -43,13 +44,17 @@ def elaborate(self, platform):
...
+                # By default, request the next command from the FIFO,
+                # and insert a delay after the current command
+                m.d.comb += out_fifo.r_en.eq(1)
+                m.d.sync += timer.eq(self.period_cyc - 1)
+
                 with m.Switch(out_fifo.r_data):
-                    m.d.comb += out_fifo.r_en.eq(1)
                     # remote_bitbang_write(int tck, int tms, int tdi)
                     with m.Case(*b"01234567"):
                         m.d.sync += Cat(bus.tdi, bus.tms, bus.tck).eq(out_fifo.r_data[:3])
```
Comment (whitequark) [inline software/glasgow/applet/interface/jtag_openocd/__init__.py:60]: Only need to set the delay to `period_cyc` here...

Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py:60]: fair point

File: software/glasgow/applet/interface/jtag_openocd/__init__.py (hunk 3 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py:84]: It's unnecessary, I think I made an editing error

File: software/glasgow/applet/interface/jtag_openocd/__init__.py (hunk 2 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py]: done

File: software/glasgow/applet/interface/jtag_openocd/__init__.py (hunk 3 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py]: removed this line

File: software/glasgow/applet/interface/jtag_openocd/__init__.py (hunk 4 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py:56]: done

File: software/glasgow/applet/interface/jtag_openocd/__init__.py [hunk 5]
```diff
@@ -64,13 +69,19 @@ def elaborate(self, platform):
...
+                    with m.Case(*b"Z"):
+                        m.d.sync += timer.eq(1000 * self.us_cyc - 1)
+                    with m.Case(*b"z"):
+                        m.d.sync += timer.eq(self.us_cyc - 1)
                     # remote_bitbang_quit(void)
                     with m.Case(*b"Q"):
                         pass
                     with m.Default():
+                        # For unknown commands, deliberately hang by requesting
+                        # no more FIFO elements, and skip the delay
```
[the remark this answers is not in the document]
Comment (neuschaefer) [reply software/glasgow/applet/interface/jtag_openocd/__init__.py]: done

**Review by whitequark — APPROVED** [review]
LGTM. Thank you for collaborating on this!
